### PR TITLE
Printing out test scores in training phase for FastTree

### LINKED ISSUE #1572: Some times we may want to provide test set in addition to training and validation sets
Imagining that one day, a user prepares a training set and a validation, and she wants to train a `FastTree` (gradient boosting decision tree) with early stopping. Then, she probably also would like to know how good our early stopping is by calculating a score based on some predictions. If this is the case, she may need another data set (let's call it test set) because it's not fair to use validation set again because the training iteration number is determined by validation set via early stopping.

### PR DESCRIPTION
**Pull request by wschin**
The current `FastTree` (defined in `FastTree.cs`) has a nice framework for handling the presence of training, validation, and test sets. This PR exposes this functionality to users by

1. Extend `MLContext` defined in `TrainContext.cs`
2. Create `Test` (defined in `Test.cs` under `FastTree` project) following what we having been doing for validation set.
3. Make sure that the test data set can be accessed in `FastTree.InitializeTests()` when using `Train` or `TrainTest` commands.

Fixes #1572.

Let's see an example command of how this works.

Command: `machinelearning>dotnet bin\AnyCPU.Debug\Microsoft.ML.Console\netcoreapp2.1\MML.dll train data=breast-cancer.txt loader=text{col=Label:R4:0 col=Features:R4:1-9} valid=breast-cancer.txt test=..\TLC\Samples\breast-cancer.txt tr=frr{tf=1 iter=2} out=model.zip`

Expected outcome printed on screen:
```
Starting to train ...
train.L1=0.291333712794827
train.L2=0.483948548949904
valid.L1=0.291333712794827
valid.L2=0.483948548949904
test[0].L1=0.291333712794827
test[0].L2=0.483948548949904
train.L1=0.244367572749215
train.L2=0.399588099386731
valid.L1=0.244367572749215
valid.L2=0.399588099386731
test[0].L1=0.244367572749215
test[0].L2=0.399588099386731
```
train*/valid*/test* are scores computed using the specified training/validation/test set.

### CODE REVIEW

File: src/Microsoft.ML.Core/Prediction/TrainerInfo.cs [hunk 1]
```diff
@@ -36,12 +36,19 @@ public sealed class TrainerInfo
...
         /// that.
         /// </summary>
         public bool SupportsValidation { get; }
 
+        /// <summary>
+        /// Whether the trainer can use test set via <see cref="TrainContext.TestSet"/>. Not implementing
+        /// this interface and returning <c>false</c> from this property is an indication the trainer does not support
+        /// that.
+        /// </summary>
+        public bool SupportsTest { get; }
```
Comment (Zruty0) [inline src/Microsoft.ML.Core/Prediction/TrainerInfo.cs]: >SupportsTest [](start = 20, length = 12)

Why do you need this?

File: src/Microsoft.ML.Data/Commands/TrainCommand.cs [hunk 2]
```diff
@@ -64,6 +64,9 @@ public sealed class Arguments : DataCommand.ArgumentsBase
             [Argument(ArgumentType.AtMostOnce, IsInputFileName = true, HelpText = "The validation data file", ShortName = "valid")]
             public string ValidationFile;
 
+            [Argument(ArgumentType.AtMostOnce, IsInputFileName = true, HelpText = "The test data file", ShortName = "test")]
+            public string TestFile;
```
Comment (Zruty0) [inline src/Microsoft.ML.Data/Commands/TrainCommand.cs]: >TestFile [](start = 26, length = 8)

I'm not sure I like the idea of having a test file be an argument to train command...

Comment (Zruty0) [reply src/Microsoft.ML.Data/Commands/TrainCommand.cs]: we could have it hidden I guess

---
In reply to: [232073278](https://github.com/dotnet/machinelearning/pull/1571#discussion_r232073278) [](ancestors = 232073278)

Comment (wschin) [reply src/Microsoft.ML.Data/Commands/TrainCommand.cs]: What's your concern? The validation set can be used to tune training parameters, so it can't be used to compute test scores sometime. If user want to have a fair mechanism to observer the change of test score in the training phase, this is the only solution she can have.

File: src/Microsoft.ML.Core/Prediction/TrainerInfo.cs (hunk 1 above)
Comment (wschin) [reply src/Microsoft.ML.Core/Prediction/TrainerInfo.cs]: It's used in command line family (for example, `TrainCommand.cs`). If a trainer says it can't support test set, the command used may to produce a `RoleMappedData` for the specified test set.

File: src/Microsoft.ML.Data/Commands/TrainCommand.cs (hunk 2 above)
Comment (rogancarr) [reply src/Microsoft.ML.Data/Commands/TrainCommand.cs]: I think we can solve this better by supporting multiple validation sets, rather than giving the training algorithms access to the test set.

Comment (wschin) [reply src/Microsoft.ML.Data/Commands/TrainCommand.cs]: No. validation set is different from test sets. Training algorithm can access information in validation set but should never touch the test set.

Comment (rogancarr) [reply src/Microsoft.ML.Data/Commands/TrainCommand.cs]: Yes, exactly. One validation set for pruning, one validation set for measuring the effectiveness of the pruning strategy. The training algorithm doesn't get access to the test set.

Comment (wschin) [reply src/Microsoft.ML.Data/Commands/TrainCommand.cs]: How could the dev of an algorithm know the difference between validation[0] and validation[1]? I prefer to use the name, test set, to warn algorithms not to touch it.

Comment (Zruty0) [reply src/Microsoft.ML.Data/Commands/TrainCommand.cs]: I think the solution as presented is good enough

---
In reply to: [232358388](https://github.com/dotnet/machinelearning/pull/1571#discussion_r232358388) [](ancestors = 232358388)
